harvard-lts/mirador-pdiiif-plugin
# Progress bar and cancel

**JIRA Ticket**: [LTSVIEWER-178](https://jira.huit.harvard.edu/browse/LTSVIEWER-178)

- [mps-viewer companion PR](https://github.com/harvard-lts/mps-viewer/pull/57)

# What does this Pull Request do?

Displays a progress bar while the generation is in progress so the user can see if the file is ready. To make this more apparent the modal window now stays open unless closed. A cancel button was added to allow the user to change their mind or choose a different set of pages. 

There were some other tweaks to what is displayed or reset when the modal is opened or closed while a download is in progress to make this more user friendly.

I also updated the tests to use [Mock Service Worker](https://mswjs.io/) (aka msw). This was with a view to add tests that would be able to trigger the progress bar and provide simulated responses to the network requests that would take place. My main stumbling block was how to interface with `showSaveFilePicker`. The tests still pass with msw, and it seems to be recommended, so I've left it here.

# How should this be tested?

1. Checkout the branch for the [companion PR on mps-viewer](https://github.com/harvard-lts/mps-viewer/pull/57)
2. Restarting the the container should be enough to have the dependencies install and run webpack
3. Viewing any of the legacy PTO and attempting to download the PDF should display a progress bar.
4. **While a PDF is being generated the following should be true:**
    1. The modal should stay open
    2. The page range input should be disabled (not accepting input)
    3. The "Download" button should say "Cancel"
    4. Closing the modal and opening it again before the download completes should mean all of the above are still true
5. **Once a PDF has finished generating the following should be true:**
    1. The progress bar should remain visible
    2. The page range input should accept input again
    3. The "Cancel" button should say "Done" and be disabled (not be clickable)
    4. Closing the modal and opening again should remove the progress bar and reset the value of the page range input
6. **Cancelling (using the "Cancel" button), an in-progress PDF generation should result in the following being true:**
    1. The progress bar should disappear
    2. The page range input should accept input again
    3. The "Cancel" button should read "Download" 

**Try the above steps in both Chrome and Firefox**

# Test coverage

Yes/No: Are changes in this pull-request covered by:

- unit tests? No (tried to add tests for this, but `showSaveFilePicker` is a barrier)
- integration tests? No

# Interested parties

Tag (@ mention) interested parties
@enriquediaz
